agent/lib_composer.c

```diff
@@ -0,0 +1,233 @@
+    return;
+  }
+
+  nrl_verbosedebug(NRL_INSTRUMENT, "%s - Composer runtime API available",
+                   __func__);
+
+  result
+      = zend_eval_string(getallrawdata, &retval, "composer_getallrawdata.php");
+  if (SUCCESS != result) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s - composer_getallrawdata.php failed",
+                     __func__);
+    return;
+  }
+  if (IS_ARRAY == Z_TYPE(retval)) {
+    zend_string* package_name = NULL;
+    zval* package_version = NULL;
+    ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL(retval), package_name,
+                                  package_version) {
+      if (NULL == package_name || NULL == package_version) {
+        continue;
+      }
+      if (nr_php_is_zval_non_empty_string(package_version)) {
+        nrl_verbosedebug(NRL_INSTRUMENT, "package %s, version %s",
+                         NRSAFESTR(ZSTR_VAL(package_name)),
+                         NRSAFESTR(Z_STRVAL_P(package_version)));
+        nr_txn_add_php_package_from_source(
+            NRPRG(txn), NRSAFESTR(ZSTR_VAL(package_name)),
+            NRSAFESTR(Z_STRVAL_P(package_version)),
+            NR_PHP_PACKAGE_SOURCE_COMPOSER);
+      }
+    }
+    ZEND_HASH_FOREACH_END();
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s - installed packages is not an array",
+                     __func__);
+  }
+  zval_dtor(&retval);
+}
+
+static char* nr_execute_handle_autoload_composer_get_vendor_path(
+    const char* filename) {
+  char* vendor_path = NULL;  // result of dirname(filename)
+  char* cp = NULL;
+
+  // vendor_path = dirname(filename):
+  // 1. copy filename to vendor_path
+  vendor_path = nr_strdup(filename);
+  // 2. // find last occurence of '/' in vendor_path
+  cp = nr_strrchr(vendor_path, '/');
+  // 3. replace '/' with '\0' to get the directory path
+  *cp = '\0';
+
+  return vendor_path;
+}
+
+static bool nr_execute_handle_autoload_composer_file_exists(
+    const char* vendor_path,
+    const char* filename) {
+  char* composer_magic_file = NULL;  // vendor_path + filename
+  bool file_exists = false;
+
+  composer_magic_file = nr_formatf("%s/%s", vendor_path, filename);
+  if (0 == nr_access(composer_magic_file, F_OK | R_OK)) {
+    file_exists = true;
+  }
+  nr_free(composer_magic_file);
+  return file_exists;
+}
+
+void nr_composer_handle_autoload(const char* filename) {
+// Composer signature file"
+#define COMPOSER_MAGIC_FILE_1 "composer/autoload_real.php"
+#define COMPOSER_MAGIC_FILE_1_LEN (sizeof(COMPOSER_MAGIC_FILE_1) - 1)
+// Composer runtime API files:
+#define COMPOSER_MAGIC_FILE_2 "composer/InstalledVersions.php"
+#define COMPOSER_MAGIC_FILE_2_LEN (sizeof(COMPOSER_MAGIC_FILE_2) - 1)
+#define COMPOSER_MAGIC_FILE_3 "composer/installed.php"
+#define COMPOSER_MAGIC_FILE_3_LEN (sizeof(COMPOSER_MAGIC_FILE_3) - 1)
+  char* vendor_path = NULL;  // result of dirname(filename)
+
+  vendor_path = nr_execute_handle_autoload_composer_get_vendor_path(filename);
+  if (NULL == vendor_path) {
```

Comment:
will vendor_path ever be null since nrstrdup [here](https://github.com/newrelic/newrelic-php-agent/pull/962/files#diff-d5f4e22aae64a7055a773470d7d61e625e9d9a782162e0a2b3d0c0638ce12682R165) will always return an empty string if given a null.

Reply:
Checking for `NULL` out of habit. I'm going to leave it in just in case `nr_execute_handle_autoload_composer_get_vendor_path` ever changes in a way that will result in `NULL` being a possible return value.